src/coins.cpp

```diff
@@ -260,6 +295,196 @@ void CCoinsViewCache::BringBestAnchorIntoCache(
     assert(GetSaplingAnchorAt(currentRoot, tree));
 }
 
+void draftMMRNode(std::vector<uint32_t> &indices, 
+                  std::vector<HistoryEntry> &entries, 
+                  HistoryNode nodeData, 
+                  uint32_t h, 
+                  uint32_t peak_pos) 
+{
+    HistoryEntry newEntry = h == 0 
+        ? libzcash::NodeToEntry(nodeData) 
+        : libzcash::NewEntry(nodeData, peak_pos - (1 << h) - 1, peak_pos - 2);
+
+    indices.push_back(peak_pos - 1);
+    entries.push_back(newEntry);
+}
+
+static inline int log2i(uint32_t x) {
+    assert(x > 0);
+    return 31 - __builtin_clz(x);
+}
+
+uint32_t CCoinsViewCache::PreloadHistoryTree(uint32_t epochId, bool extra, std::vector<HistoryEntry> &entries, std::vector<uint32_t> &entry_indices) {
```

Comment:
It does not fully correspond to get_peaks

As we discussed before, this preloads some peaks expected to be touched by subsequent update (append/delete), so that we don't query database while generating new root during that operation

Reply:
Ah yes. This should be documented so that it's clearer from the source code what is going on (e.g. how "expected peaks" are picked), and ideally the ZIP would also be updated to describe this optimisation.

Reply:
Unresolving this conversation because I don't think the ZIP has been updated.

Reply:
Yes I'd also like to see an explanation in the ZIP.